src/model/document.ts

```diff
@@ -56,110 +54,96 @@ export class ScannedDocument {
-    this.isInline = base.isInline;
-    this.parsedDocument = base.document;
-    this.sourceRange = base.sourceRange;
-    this._warnings = base.warnings;
-
-    if (base.isInline) {
-      this.identifiers = new Set();
-    } else {
-      this.identifiers = new Set([this.url]);
+    if (analyzer == null) {
+      throw new Error('analyzer is null');
     }
-    this.kinds = new Set(['document', `${this.parsedDocument.type}-document`]);
-    this._addFeature(this);
-  }
+    this._scannedDocument = base;
+    this.analyzer = analyzer;
 
-  /**
-   * To handle recursive dependency graphs we must track whether we've started
-   * resolving this Document so that we can reliably early exit even if one
-   * of our dependencies tries to resolve this document.
-   */
-  private _begunResolving = false;
-  private _resolve(base: ScannedDocument) {
-    if (this._begunResolving) {
-      return;
+    if (!base.isInline) {
+      this.identifiers.add(this.url);
     }
-    this._begunResolving = true;
-    for (const scannedFeature of base.features) {
-      if (scannedFeature instanceof ScannedImport) {
-        this._resolveScannedImport(scannedFeature);
+    this.kinds.add(`${this.parsedDocument.type}-document`);
+    // this._resolve();
+  }
 
-      } else if (scannedFeature instanceof InlineParsedDocument) {
-        this._resolveInlineDocument(scannedFeature);
+  get url(): string {
+    return this._scannedDocument.url;
+  }
 
-      } else if (isResolvable(scannedFeature)) {
-        const feature = scannedFeature.resolve(this._rootDocument);
-        this._addFeature(feature);
-      }
-    }
-    this._doneResolving = true;
+  get isInline(): boolean {
+    return this._scannedDocument.isInline;
   }
 
-  private _resolveScannedImport(scannedImport: ScannedImport) {
-    const imprt = scannedImport.resolve(this._rootDocument);
-    this._addFeature(imprt);
+  get parsedDocument(): ParsedDocument<any, any> {
+    return this._scannedDocument.document;
+  }
 
-    const scannedDoc = scannedImport.scannedDocument;
-    if (!scannedDoc) {
-      // There was a load or parse error, the scanned doc doesn't exist.
-      return;
-    }
+  get sourceRange(): SourceRange {
+    return this._scannedDocument.sourceRange;
+  }
 
-    const document = new Document(scannedDoc, this._rootDocument);
-    imprt.document = document;
-    this._addFeature(document);
+  get _warnings(): Warning[] {
+    return this._scannedDocument.warnings;
+  }
 
-    document._resolve(scannedDoc);
+  get resolved(): boolean {
+    return this._doneResolving;
   }
 
-  private _resolveInlineDocument(inlineDoc: InlineParsedDocument) {
-    if (!inlineDoc.scannedDocument) {
-      // Parse error on the inline document.
+  get type(): string {
+    return this.parsedDocument.type;
+  }
+
+  /**
+   * Resolves all features of this document, so that they have references to all
+   * their dependencies.
+   *
+   * This method can only be called once
+   */
+  resolve() {
```

Comment:
It's called outside of this class now. I'd like to make ScannedDoument a resolvable and do this work in its `resolve()`